include/net/http_client.h

```diff
@@ -0,0 +1,291 @@
+#endif
+
+struct http_ctx;
+
+/* HTTP header fields struct */
+struct http_field_value {
+	/** Field name, this variable will point to the beginning of the string
+	 *  containing the HTTP field name
+	 */
+	const char *key;
+
+	/** Value, this variable will point to the beginning of the string
+	 *  containing the field value
+	 */
+	const char *value;
+
+	/** Length of the field name */
+	u16_t key_len;
+
+	/** Length of the field value */
+	u16_t value_len;
+};
+
+/* Is there more data to come */
+enum http_final_call {
+	HTTP_DATA_MORE = 0,
+	HTTP_DATA_FINAL = 1,
+};
+
+/* http header filed and value */
+enum http_header_field_val {
+	HTTP_HEADER_FIELD = 0,
+	HTTP_HEADER_VALUE = 1,
+};
+
+/**
+ * @typedef http_response_cb_t
+ * @brief Callback used when a response has been received from peer.
+ *
+ * @param ctx HTTP context.
+ * @param data Received body point
+ * @param received body length
+ * @param final_data Does this data buffer contain all the data or
+ * is there still more data to come.
+ * @param user_data A valid pointer on some user data or NULL
+ */
+typedef void (*http_response_cb_t)(struct http_ctx *ctx,
+					const char *body,
+					int body_len,
+					enum http_final_call final_data);
+
+/**
+ * @typedef http_get_filed_value_cb_t
+ * @brief Callback used when got http header filed and value.
+ *
+ * @param ctx HTTP context.
+ * @param data of header
+ * @param header field or value length
+ * @param http_header_field_val indicate it's filed or value
+ */
+typedef void (*http_get_filed_value_cb_t)(struct http_ctx *ctx,
+					const char *at,
+					int at_len,
+					enum http_header_field_val f_v);
+/**
+ * HTTP client request. This contains all the data that is needed when doing
+ * a HTTP request.
+ */
+struct http_request {
+	/** The HTTP method: GET, HEAD, OPTIONS, POST, ... */
+	const char *method_str;
+
+	/** The URL for this request, for example: /index.html */
+	const char *path;
+
+	/** The HTTP header with optional*/
+	const char *header_fields;
+
+	/** The size of HTTP header fields */
+	const u16_t header_fields_size;
+
+	/** http header keep-alive*/
+	const bool keep_alive;
+
+};
+
+struct http_resp {
+	/** HTTP Content-Length field value */
+	size_t content_length;
+
+	/** Content length parsed. This should be the same as
+	 * the content_length field if parsing was ok.
+	 */
+	size_t processed;
+
+	/* https://tools.ietf.org/html/rfc7230#section-3.1.2
+	 * The status-code element is a 3-digit integer code
+	 *
+	 * The reason-phrase element exists for the sole
+	 * purpose of providing a textual description
+	 * associated with the numeric status code. A client
+	 * SHOULD ignore the reason-phrase content.
+	 */
+	char http_status[HTTP_STATUS_STR_SIZE];
+
+	u8_t cl_present;
+	u8_t body_found;
+	u8_t message_complete;
+};
+
+/**
+ * Http context information. This contains all the data that is
+ * needed when working with http API.
+ */
+struct http_ctx {
+
+	/** socket id with http conenct*/
+	u16_t sock_id;
```

Comment:
you could move that attributes elsewhere so it would make a better structure alignment.

Reply:
as you comment, is this structure(http_ctx)'s elsewhere or other structure?

Reply:
in this structure yes